### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by AlexTMallen**
- integrate wandb
- separate out some hyperparameters for weak and strong
- use kl loss for w2s for interpretability (just xent plus a constant)
- avoid subprocess
- use torch for saving model checkpoints
- fix batching generator

### CODE REVIEW

File: train_simple.py [hunk 1]
```diff
@@ -46,9 +49,9 @@ def main(
     sweep_subfolder: str = "default",
```
Comment (ojh31) [inline train_simple.py:49]: Seems odd to have "sweep" in a parameter name here when this is not the sweep script

File: weak_to_strong/train.py [hunk 2]
```diff
@@ -10,29 +10,40 @@
...
 
 import weak_to_strong.logger as logger
 from weak_to_strong.common import clear_mem, to_batch
 from weak_to_strong.eval import eval_model_acc
-from weak_to_strong.loss import xent_loss
+from weak_to_strong.loss import kl_loss
 from weak_to_strong.model import TransformerWithHead
 from weak_to_strong.config import ModelConfig
 
+def save(model: torch.nn.Module, save_path: str, optimizer=None, scheduler=None):
```
Comment (ojh31) [inline weak_to_strong/train.py:22]: I think this function could be cleaned up a bit, removing redundant comments, preceding with 2 newlines and avoid re-making the path for the print statment, perhaps by moving inside `save_torch`?

File: weak_to_strong/train.py [hunk 3]
```diff
@@ -150,20 +165,21 @@ def lr_schedule_fn(step):
 
             try:
                 auroc = roc_auc_score(all_hard_labels.cpu(), all_logprobs.cpu())
-            except ValueError:
+            except ValueError as e:
```
Comment (ojh31) [inline weak_to_strong/train.py:168]: Nice!

**Review by ojh31 — APPROVED** [review]
Some nice improvements here, in particular I'm a big fan of wandb. I left a few minor comments mostly to aid readability for others.

File: train_simple.py [hunk 4]
```diff
@@ -23,13 +24,15 @@ def main(
     batch_size: int = 32,
     max_ctx: int = 1024,
     ds_name: str = "sciq",
-    loss: str = "xent",
-    n_docs: int = 20000,
+    loss: str = "kl",
+    n_train1_docs: int = 20000,
```
[the remark this answers is not in the document]
Comment (AlexTMallen) [reply train_simple.py:28]: Number of documents (examples), yeah I could document this better

File: train_simple.py [hunk 5]
```diff
@@ -46,9 +49,9 @@ def main(
     sweep_subfolder: str = "default",
     # Set to a very large value so that by default we don't do any intermediate evals but
     # still do final evals (which requires eval_every to be set to a non-zero, non-None value)
-    eval_every: int = 1000000,
+    strong_eval_every: int = 10000000,
```
[the remark this answers is not in the document]
Comment (AlexTMallen) [reply train_simple.py:52]: This wasn't a particularly strong choice, just that I'm not super interested in seeing the learning curve for finetuning on ground truth so it goes with a single eval at the end. Also, now that I think of it, "weak" and "strong" are probably not the right words, because we also do ground-truth finetuning on strong, so I should rename these to "gt_..." and "w2s_..."

File: train_simple.py (hunk 1 above)
Comment (AlexTMallen) [reply train_simple.py:49]: This was OpenAI's choice, it's passing information about which sweep this run is a part of to this script so that it can keep track of where to store results

File: train_simple.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ojh31) [reply train_simple.py:52]: Ah yes that makes more sense, thanks!
